Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs

```diff
@@ -0,0 +1,299 @@
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Http;
+using System.IO;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>
+    /// Credential for authorizing calls using OAuth 2.0.
+    /// It is a convenience wrapper that allows handling of different types of 
+    /// credentials (like ServiceAccountCredential, ComputeCredential or UserCredential)
+    /// in a unified way.
+    /// <para>
+    /// See <see cref="GetApplicationDefaultCredential"/>.
+    /// </para>
+    /// </summary>
+    public abstract class GoogleCredential : IConfigurableHttpClientInitializer, ITokenAccess
+    {
+        private static DefaultCredentialProvider defaultCredentialProvider = new DefaultCredentialProvider();
+
+        /// <summary>
+        /// <para>Returns the Application Default Credentials which are ambient credentials that identify and authorize 
+        /// the whole application.</para>
+        /// <para>The ambient credentials are determined as following order:</para>
+        /// <list type="number">
+        /// <item> 
+        /// <description>The environment variable GOOGLE_APPLICATION_CREDENTIALS is checked. If this variable is specified, it 
+        /// should point to a file that defines the credentials. The simplest way to get a credential for this purpose is to 
+        /// create a service account using the <a href="https://console.developers.google.com">Google Developers Console</a> in 
+        /// the section APIs &amp; Auth, in the sub-section Credentials. Create a service account or choose an existing one and 
+        /// select Generate new JSON key. Set the environment variable to the path of the JSON file downloaded.</description> 
+        /// </item> 
+        /// <item> 
+        /// <description>If you have installed the Google Cloud SDK on your machine and have run the command 
+        /// <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>, your identity can be used as 
+        /// a proxy to test code calling APIs from that machine.</description> 
+        /// </item> 
+        /// <item> 
+        /// <description>If you are running in Google Compute Engine production, the built-in service account associated with the 
+        /// virtual machine instance will be used.</description> 
+        /// </item> 
+        /// </list>
+        /// </summary>
+        public static Task<GoogleCredential> GetApplicationDefaultAsync()
+        {
+            return defaultCredentialProvider.GetDefaultCredentialAsync();
+        }
+
+        /// <summary>
+        /// Load credential from stream containing JSON credential data.
+        /// <para>
+        /// The stream can contain a Service Account key file in JSON format from the Google Developers
+        /// Console or a stored user credential using the format supported by the Cloud SDK.
+        /// </para>
+        /// </summary>
+        public static GoogleCredential FromStream(Stream stream)
```

Comment:
This actually comes handy because it allows you to read JSON service account credential  or JSON user credential from a file explicitly. 
Imagine you download the JSON key file from cloud console and want to use it with your app. If you used ServiceAccountCredential directly, you would need to parse the JSON first and extract values of the fields and initialize service account credential accordingly. 
This method can do all of that.